crates/sb_core/js/01_http.js

```diff
@@ -0,0 +1,554 @@
+} from "ext:deno_web/06_streams.js";
+import { serve } from "ext:deno_http/00_serve.js";
+import { SymbolDispose } from "ext:deno_web/00_infra.js";
+
+const connErrorSymbol = Symbol("connError");
+
+/** @type {(self: HttpConn, rid: number) => boolean} */
+let deleteManagedResource;
+
+class HttpConn {
+  #rid = 0;
+  #closed = false;
+  #remoteAddr;
+  #localAddr;
+
+  // This set holds resource ids of resources
+  // that were created during lifecycle of this request.
+  // When the connection is closed these resources should be closed
+  // as well.
+  #managedResources = new SafeSet();
+
+  static {
+    deleteManagedResource = (self, rid) =>
+      SetPrototypeDelete(self.#managedResources, rid);
+  }
+
+  constructor(rid, remoteAddr, localAddr) {
+    this.#rid = rid;
+    this.#remoteAddr = remoteAddr;
+    this.#localAddr = localAddr;
+  }
+
+  /** @returns {number} */
+  get rid() {
+    return this.#rid;
+  }
+
+  /** @returns {Promise<RequestEvent | null>} */
+  async nextRequest() {
+    let nextRequest;
+    try {
+      nextRequest = await op_http_accept(this.#rid);
+    } catch (error) {
+      this.close();
+      // A connection error seen here would cause disrupted responses to throw
+      // a generic `BadResource` error. Instead store this error and replace
+      // those with it.
+      this[connErrorSymbol] = error;
+      if (
+        ObjectPrototypeIsPrototypeOf(BadResourcePrototype, error) ||
+        ObjectPrototypeIsPrototypeOf(InterruptedPrototype, error) ||
+        StringPrototypeIncludes(error.message, "connection closed")
+      ) {
+        return null;
+      }
+      throw error;
+    }
+    if (nextRequest == null) {
+      // Work-around for servers (deno_std/http in particular) that call
+      // `nextRequest()` before upgrading a previous request which has a
+      // `connection: upgrade` header.
+      await null;
+
+      this.close();
+      return null;
+    }
+
+    const { 0: streamRid, 1: method, 2: url } = nextRequest;
+    SetPrototypeAdd(this.#managedResources, streamRid);
+
+    /** @type {ReadableStream<Uint8Array> | undefined} */
+    let body = null;
+    // There might be a body, but we don't expose it for GET/HEAD requests.
+    // It will be closed automatically once the request has been handled and
+    // the response has been sent.
+    if (method !== "GET" && method !== "HEAD") {
+      body = readableStreamForRid(streamRid, false);
+    }
+
+    const innerRequest = newInnerRequest(
+      method,
+      url,
+      () => op_http_headers(streamRid),
+      body !== null ? new InnerBody(body) : null,
+      false,
+    );
+
+    const abortController = new AbortController();
+    const request = fromInnerRequest(
+      innerRequest,
+      abortController.signal,
+      "immutable",
+      false,
+    );
+
+    const respondWith = createRespondWith(
+      this,
+      streamRid,
+      abortController,
+    );
+
+    return { request, respondWith, streamRid };
```

Comment:
Note: This is the only purpose this file exists for.
We need `streamRid` for protocol upgrades, but Deno hasn't exposed it at all.